src/query/api/v1/handler/prom/read.go

```diff
@@ -159,6 +135,43 @@ func newReadHandler(
 	return handler, nil
 }
 
+type queryShadowing struct {
+	// This URL doesn't includes the path, "api/v1/query_range" or "api/v1/query".
+	// It shouldn't end with a slash('/').
+	shadowQueryURL string
+	workerPool     xsync.WorkerPool
+	client         *http.Client
+	failedQueryCounter tally.Counter
+	respondedQueryCounter tally.Counter
+	responded2xxQueryCounter tally.Counter
+	skippedQueryCounter tally.Counter
+}
+
+func getHttpClient() *http.Client {
+	t := http.DefaultTransport.(*http.Transport).Clone()
+	t.MaxIdleConns = 10
+	t.MaxConnsPerHost = 10
+	t.MaxIdleConnsPerHost = 10
+	return &http.Client{
+		Timeout:   4 * 60 * time.Second,
+		Transport: t,
+	}
+}
+
+func newQueryShadowing(shadowQueryURL string, numWorkers int, scope tally.Scope) *queryShadowing {
+	workerPool := xsync.NewWorkerPool(numWorkers)
+	workerPool.Init()
+	return &queryShadowing{
+		shadowQueryURL: shadowQueryURL,
+		workerPool:     workerPool,
+		client:         getHttpClient(),
+		failedQueryCounter: scope.Counter("failed_shadow_query"),
```

Comment:
I think it is still valuable to measure latency even though it is not apple to apple compare, we just can measure how Pantheon performance against those ad-hoc queries. It won't cost much to gain extra visibility.